client/keycloakJWTReceiverCachedInMemory.go

```diff
@@ -0,0 +1,68 @@
+// SPDX-FileCopyrightText: 2024 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+package client
+
+import (
+	"fmt"
+
+	"github.com/Nerzal/gocloak/v12"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/rs/zerolog/log"
+)
+
+type KeycloakJWTReceiverCachedInMemory struct {
+	keycloakRepository IKeycloakRepository
+	cachedToken        *gocloak.JWT
+}
+
+func NewKeycloakJWTReceiverCachedInMemory(keycloakRepository IKeycloakRepository) *KeycloakJWTReceiverCachedInMemory {
+	return &KeycloakJWTReceiverCachedInMemory{
+		keycloakRepository: keycloakRepository,
+	}
+}
+
+func isTokenValid(token *gocloak.JWT) bool {
+	if token == nil {
+		return false
+	}
+
+	parser := jwt.NewParser()
+	claims := &jwt.MapClaims{}
+
+	_, _, err := parser.ParseUnverified(token.AccessToken, claims)
+	if err != nil {
+		log.Error().Msgf("couldn't parse JWT access token: %v", err)
+		return false
+	}
+
+	err = claims.Valid()
+	if err != nil {
+		log.Debug().Msgf("JWT access token is invalid: %v", err)
+		return false
+	}
+
+	return true
+}
+
+func (k *KeycloakJWTReceiverCachedInMemory) getClientToken(clientName, clientSecret string) (*gocloak.JWT, error) {
+	if k.cachedToken == nil || !isTokenValid(k.cachedToken) {
```

Comment:
i think we should add a mutex here - there can be race conditions as several goroutines try to access opensearch and therefore maybe request a token and chachedToken seems to be a shared resource